### PR TITLE
feat: Implement a k-d tree data structure

### PR DESCRIPTION
**Pull request by stephenswat**
This pull request is part three of four in my attempt to implement an orthogonal range search seed finder. Parts one and two are #985 and #986.

_k_-d trees are data structures that allow certain operations to be performed efficiently on point cloud data. Most commonly, these
operations are (orthogonal) range search and nearest neighbour search. For my implementation, I need only the first, and as such nearest neighbour search is omitted in name of the YAGNI principle.

I have implemented this class with performance and GPU-compatibility in mind, and therefore the memory is efficiently laid out wherever possible (although some improvements can still be made later).

The implementation supports a more classical C++ interface where vectors of items are returned, as well as a more functional approach which employs higher-order functions to effect the required computation. This functional approach is the backbone to the classical approach as well.

I believe that a _k_-d tree implementation is a good thing to have in Acts, since they are very versatile and I can see them being used in several other parts of the code.

Again, I have added tests for the k-d tree as well, to ensure that the implementation works as intended.

This pull request cannot be merged until #985 and #986 are merged.

### CODE REVIEW

**Comment by stephenswat** [comment]
Hmm, that test coverage is quite low. I'll work on improving that.

**Comment by stephenswat** [comment]
I've updated the commit with the following changes:

1. The `str` methods have been renamed to `toString` methods, just like in #986.
2. The license header has been fixed.
3. An undocumented parameter which was causing the docs generation to crash has been documented.
4. Additional tests have been added to cover some edge cases.
5. The k-d tree has been modified to work with the new half-open intervals (which was somewhat awkward).

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+  ///
+  /// This method allows the user more control in where the result is written
+  /// to.
+  ///
+  /// @tparam OutputIt The type of the output iterator.
+  ///
+  /// @param r The range to search for.
+  /// @param i The iterator to write the output to.
+  template <typename OutputIt>
+  void rangeSearchInserter(const range_t &r, OutputIt i) const {
```
Comment (asalzburger) [inline Core/include/Acts/Utilities/KDTree.hpp:160]: for my understanding: does this create a local copy of `i` within the method?

File: Core/include/Acts/Utilities/KDTree.hpp [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+    /// @param _b The iterator for the start of the range of elements.
+    /// @param _e The iterator for the end of the range of elements.
+    KDTreeNode(iterator_t _b, iterator_t _e) : KDTreeAbstractNode(_b, _e) {
+      // This constant determines the maximum number of elements where we still
+      // calculate the exact median of the values for the purposes of
+      // splitting. In general, the closer the pivot value is to the true
+      // median, the more balanced the tree will be. However, calculating the
+      // median exactly is an O(n log n) operation, while approximating it is
+      // an O(1) time.
+      constexpr std::size_t max_exact_median = 128;
```
Comment (asalzburger) [inline Core/include/Acts/Utilities/KDTree.hpp:404]: Where does the number 128 come from - did you time that this is a good cross-over point?

**Review by asalzburger — APPROVED** [review]
Very nice to get this together, I saw you fixed the `doc` compilation already, so that's great.

I had two questions for my understanding, but other than that I am happy to merge this in.

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 1 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/KDTree.hpp:160]: Yes. As far as I know passing iterators by value is the canonical way to do that in C++. This sort of makes sense if you think about it, because you have, in essence, four options for passing things to C++ functions:

1. As values, which I've done here.
2. As lvalue references, which would not allow you pass any temporaries to the function, so that would be inconvenient.
3. As const lvalue references, which allow you to pass temporaries but would not allow you to write to the iterator, so that would be useless as well.
4. As rvalue references, which in this case is pretty much the same as passing by value, but with extra steps in this case. Iterators are designed to be super-low overhead, so moving them doesn't really make sense for them.

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 2 above)
Comment (stephenswat) [reply Core/include/Acts/Utilities/KDTree.hpp:404]: So this is a very good question with a lot of backstory behind it. When you traverse a binary tree, the number of steps depends on how well balanced it is. In the best case, if your tree with _n_ nodes is perfectly balanced, it takes _O_(log _n_) steps, and in the worst case, if your tree is perfectly unbalanced (we call this a degenerate binary tree), then it takes _O_(_n_) steps. From this it follows that you would like your tree to be as balanced as possible.

However, creating a balanced tree is not always easy. In order to generate a perfectly balanced tree, you need to split each subrange exactly in the middle. This is why k-d trees are usually built using the median, because of course the median, by definition, splits the range in half. If you always pick the exact median, you will always end up with a well-balanced tree.

Now the problem is that selecting the median value is an _O_(_n_ log _n_) operation. This means that you're trading off time spent on constructing on time spent looking up values in the tree. And this is where the problem arises: the optimal cut-off point where.

Luckily, in the particular use case I am targeting, the values in most of the dimension are symmetrically distributed (I think); the phi angle is uniformly distributed, and the z axis is probably mostly symmetrical. For symmetrical distribution the mean value and the median value are of course pretty close. The radius is not symmetrical, and there could be a bit of an issue there. But the mean approach should work fairly well for this particular use case.

There is a third solution, which is the median-of-median strategy, where you approximate the median of a range of values by finding the median of medians of sub-ranges. However, I didn't think this would be noticeable enough to be worth implementing.

So, to summarize, there is a Pareto optimal front of where you should cut off your exact median computation, but it depends on a lot of things: how big is your data set; how many times are you accessing this data per tree constructed; how symmetric is the distribution of your data; how large are your search ranges; what is the computational cost of sorting on your platform; and quite a few more.

The value of 128 here was chosen arbitrarily here. I picked quite a low value, because as I mentioned the data I am working with is relatively symmetric, and therefore the performance loss of using an approximate median is a lot lower than it is for arbitrary data.

I was thinking of making this parameter a compile-time template or even a run-time configuration parameter, but I didn't think it would be worth it for now. I am happy to implement that though, if you'd like!

**Comment by stephenswat** [comment]
So I've noticed the Mac OS CI build timed out; is that something that sometimes happens? Is the OS X build just slow?

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 1 above)
Comment (asalzburger) [reply Core/include/Acts/Utilities/KDTree.hpp:160]: Thanks!

File: Core/include/Acts/Utilities/KDTree.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/Utilities/KDTree.hpp:404]: Excellent answer, thank you - I learned something!

**Comment by asalzburger** [comment]
> So I've noticed the Mac OS CI build timed out; is that something that sometimes happens? Is the OS X build just slow?

Ha,  no idea, let's just re-start the CI and see.
